File: server/app/utils/date_utils.py
from datetime import datetime, timedelta

def parse_time_to_minutes(time_str: str) -> int:
    parts = time_str.split(":")
    return int(parts[0]) * 60 + int(parts[1])

def format_minutes_to_time(total_minutes: int) -> str:
    hours = total_minutes // 60
    mins = total_minutes % 60
    return f"{hours:02d}:{mins:02d}"

def get_day_name(date_str: str) -> str:
    dt = datetime.strptime(date_str, "%Y-%m-%d")
    return dt.strftime("%A")

def is_date_in_range(date_str: str, start_date_str: str, end_date_str: str) -> bool:
    target = datetime.strptime(date_str, "%Y-%m-%d")
    start = datetime.strptime(start_date_str, "%Y-%m-%d")
    end = datetime.strptime(end_date_str, "%Y-%m-%d")
    return start <= target <= end

def generate_discrete_slots(
    working_start: str = "09:00",
    working_end: str = "17:00",
    break_start: str = "13:00",
    break_end: str = "14:00",
    slot_duration_minutes: int = 30
):
    w_start = parse_time_to_minutes(working_start)
    w_end = parse_time_to_minutes(working_end)
    b_start = parse_time_to_minutes(break_start) if break_start else None
    b_end = parse_time_to_minutes(break_end) if break_end else None

    slots = []
    curr = w_start

    while curr + slot_duration_minutes <= w_end:
        slot_end = curr + slot_duration_minutes
        
        # Skip break times
        if b_start is not None and b_end is not None:
            if not (slot_end <= b_start or curr >= b_end):
                curr = b_end
                continue

        slots.append({
            "startTime": format_minutes_to_time(curr),
            "endTime": format_minutes_to_time(slot_end)
        })
        curr = slot_end

    return slots

File: server/src/utils/dateUtils.ts
/**
 * Formats a Date object to YYYY-MM-DD
 */
export function formatDate(date: Date): string {
  const y = date.getFullYear();
  const m = String(date.getMonth() + 1).padStart(2, '0');
  const d = String(date.getDate()).padStart(2, '0');
  return `${y}-${m}-${d}`;
}

/**
 * Returns the day of the week in English (e.g., "Monday")
 */
export function getDayName(dateString: string): string {
  const [year, month, day] = dateString.split('-').map(Number);
  const date = new Date(year, month - 1, day);
  return date.toLocaleDateString('en-US', { weekday: 'long' });
}

/**
 * Adds minutes to an HH:mm string and returns new HH:mm
 */
export function addMinutesToTime(timeStr: string, minutesToAdd: number): string {
  const [h, m] = timeStr.split(':').map(Number);
  const totalMinutes = h * 60 + m + minutesToAdd;
  const newH = Math.floor(totalMinutes / 60);
  const newM = totalMinutes % 60;
  return `${String(newH).padStart(2, '0')}:${String(newM).padStart(2, '0')}`;
}

/**
 * Checks if time A is strictly before time B (HH:mm format)
 */
export function isTimeBefore(timeA: string, timeB: string): boolean {
  return timeA < timeB;
}

/**
 * Checks if a slot falls within a break range
 */
export function isSlotDuringBreak(
  slotStart: string,
  slotEnd: string,
  breakStart?: string | null,
  breakEnd?: string | null
): boolean {
  if (!breakStart || !breakEnd) return false;
  // Overlaps if slotStart < breakEnd and slotEnd > breakStart
  return slotStart < breakEnd && slotEnd > breakStart;
}

/**
 * Generates all discrete time slots for a doctor's working schedule
 */
export function generateDoctorSlots(
  startHour: string,
  endHour: string,
  durationMinutes: number,
  breakStart?: string | null,
  breakEnd?: string | null
): Array<{ startTime: string; endTime: string }> {
  const slots: Array<{ startTime: string; endTime: string }> = [];
  let current = startHour;

  while (isTimeBefore(current, endHour)) {
    const next = addMinutesToTime(current, durationMinutes);
    if (next > endHour) break;

    // Check if slot falls in break
    if (!isSlotDuringBreak(current, next, breakStart, breakEnd)) {
      slots.push({ startTime: current, endTime: next });
    }
    current = next;
  }

  return slots;
}

/**
 * Checks if a date string is between startDate and endDate inclusive
 */
export function isDateInRange(targetDate: string, startDate: string, endDate: string): boolean {
  return targetDate >= startDate && targetDate <= endDate;
}
